lib/util/ldaputil.js

```diff
@@ -76,29 +76,135 @@ export const login = function(options, username, password) {
+            if (err) {
+                resolver.resolve([])
+                return
+            }
+
+            const groups = []
+            search.on('searchEntry', entry => groups.push(entry.object.dn))
+            search.on('end', () => resolver.resolve(groups))
+            search.on('error', () => resolver.resolve(groups))
+
+            resolver.promise.finally(function() {
+                search.removeAllListeners('searchEntry')
+                search.removeAllListeners('end')
+                search.removeAllListeners('error')
+            })
+        })
+
+        return resolver.promise
+    }
     function tryBind(client, entry) {
         return new Promise(function(resolve, reject) {
-            client.bind(entry.object.dn, password, function(err) {
+            client.bind(entry.object.dn, password, async(err) => {
                 if (err) {
                     reject(new InvalidCredentialsError(username))
+                    return
                 }
-                else {
+
+                try {
+                    const groups = await findUserGroups(client, entry.object.dn)
+                    resolve({...entry.object, memberOf: groups})
+                }
+                catch (err) {
                     resolve(entry.object)
                 }
             })
         })
     }
-    return tryConnect().then(function(client) {
-        return tryFind(client)
-            .then(function(entry) {
-                return tryBind(client, entry)
-            })
-            .finally(function() {
-                client.unbind()
-            })
-    })
+    return tryConnect().then((client) => tryFind(client)
+        .then(function(entry) {
+            return tryBind(client, entry)
+        })
+        .then(function(userWithGroups) {
+            if (!userWithGroups.memberOf && userWithGroups.groups) {
+                userWithGroups.memberOf = userWithGroups.groups
+                delete userWithGroups.groups
+            }
+            return userWithGroups
+        })
+        .finally(function() {
+            client.unbind()
+        }))
 }
 export const email = function(user) {
-    return user.mail || user.email || user.userPrincipalName
+    const userEmail = user.mail || user.email || user.userPrincipalName
+    if (userEmail) {
+        return userEmail
+    }
+
+    // If an LDAP user does not have a mail, we try to make one up.
+    // A unique email must be specified for all users.
+    let username
+
+    if (user.uid) {
+        username = Array.isArray(user.uid) ? user.uid[user.uid.length - 1] : user.uid
+    }
+    else if (user.cn) {
+        username = Array.isArray(user.cn) ? user.cn[user.cn.length - 1] : user.cn
+    }
+    else {
+        const dnMatch = (user.dn || '').match(/cn=([^,]+)/i)
+        username = dnMatch ? dnMatch[1] : Math.random().toString(20).substring(2, 8)
+    }
+
+    return `${username}@ldap.ru`
```

Comment:
Let's use ldap.local or an fqdn of the ldap server?
Also show a warning in logs that this is not the recommended behaviour